### ISSUE TITLE
Inconsistent semantics of cheri_no_provenance between binary and unary operations

### ISSUE THREAD

**Issue by acoplan-arm** [CTSRD-CHERI/llvm-project#542]
Consider the following functions:

```
__intcap_t f(__intcap_t c __attribute((cheri_no_provenance))) { return c + 1; }
__intcap_t g(__intcap_t c __attribute((cheri_no_provenance))) { return -c; }
__intcap_t h(__intcap_t c __attribute((cheri_no_provenance))) { return ~c; }
```

on purecap Morello, for `f`, we generate:
```
f:                                      // @f
        gcvalue x8, c0
        add     x8, x8, #1                      // =1
        mov     x0, #0
        add     c0, c0, x8, uxtx
        ret     c30
```
i.e. we use a null-derived capability, since the argument is marked with the `cheri_no_provenance` attribute. However, for the unary operations in `g` and `h`, we carry the provenance from the argument instead:

```
g:                                      // @g
        gcvalue x8, c0
        neg     x8, x8
        scvalue c0, c0, x8
        ret     c30
h:                                      // @h
        gcvalue x8, c0
        mvn     x8, x8
        scvalue c0, c0, x8
        ret     c30
```

which seems inconsistent. It seems that we should use a null-derived capability in all of these cases.

**Comment by jrtc27** [comment]
The attribute is not a "please remove provenance" attribute, it's an "I promise this has no provenance" (or at least "I promise this has no provenance, or if it does I don't care about whether it's preserved") attribute, and exists solely to resolve ambiguity in binary operators. For binary operators that means if both are annotated as not having provenance then we null derive in order to provide determinism. For unary operators we don't do anything as there is no ambiguity with respect to the provenance. We could choose to optimise it to null-derive the result, but I see that more of an optimisation than correctness (though on Morello the lack of being able to specify czr in add instructions means doing so would bloat the code, compared with CHERI-MIPS and CHERI-RISC-V where there are no such issues, though with sufficiently clever (and careful, to ensure it doesn't reintroduce provenance) peephole optimisation you can exploit the merged register file and omit the null rederivation, just using the super-register of the integer register). There are always going to be ways where such attributed types have their provenance preserved; we're not going to rederive from null on every assignment, function call, value return, etc.

That is, if your intent is to _remove_ provenance then you _must_ cast to `ptraddr_t` and back to do so.

**Comment by acoplan-arm** [comment]
> we're not going to rederive from null on every assignment, function call, value return, etc.

Yeah, that makes sense. I suppose the question is then where you draw the line. One possibility is that you null derive on any operation that results in updating the address value of the capability where all of the intcap operands are marked with `cheri_no_provenance` attribute, which is what I had in mind. @arichardson seemed to think this would be better from a consistency POV as well.

It would be nice if we could agree on a clear semantics for the attribute for consistency between implementations.

**Comment by arichardson** [comment]
I agree with @jrtc27 that this is not a correctness issue, these attributes were not intended to remove provenance but to indicate that the capability metadata is *not required* to be preserved.  If you want to remove provenance, casting to a type such as `ptraddr_t` is the right approach.

However, I do think that we should generate the same code for the following two functions:
```
__intcap_t unary(__intcap_t c __attribute((cheri_no_provenance))) { return -c; }
__intcap_t binary(__intcap_t c __attribute((cheri_no_provenance))) { return 0 - c; }
```

The only reason I didn't do this originally is that the ambiguous case only happens for binary operators so I didn't think about changing the code generation for unary ones. Adjusting them to also handle the attribute should be trivial.

I also didn't consider code that adds the attribute explicitly in much detail - in most cases the attribute is created by the compiler internally to avoid ambiguous provenance warnings. The explicit attribute only really seems useful to me if you need to maintain a structure layout (and therefore can't change the variable to ptraddr_t). Annotating a structure field can make sense since it might remove many ambiguous provenance warnings without having to change all the uses.

If you look at the AST dump for the following code (https://cheri-compiler-explorer.cl.cam.ac.uk/z/h5Kvzj) you can see that clang adds the no_provenance attribute to the cast expression to avoid warning about ambiguous provenance:
```
#include <stdint.h>

uintptr_t test(uintptr_t a) {
    long mask = 0xf;
    return a & ~(uintptr_t)mask;
}
```
There is almost certainly a better way of implementing this provenance tracking (without the cost of full dataflow analysis) but at the time it seemed like the easiest approach was to attach the `no_provenance` attribute to all uintptr_t types that are created via integer promotion/casts/ etc. However, this does mean that we still warn about ambiguous provenance for the following code:
```
#include <stdint.h>

uintptr_t test(uintptr_t a) {
    uintptr_t mask = 0xf;
    return a & ~mask;
}
```
